src/HealthGPS.Datastore/datamanager.cpp

```diff
@@ -1,86 +1,60 @@
-    }
-
-    // Strip URL prefix and load file from local filesystem
-    const auto filename = std::filesystem::path{uri.path()}.filename();
-    const auto schema_path = schema_directory / filename;
-    auto ifs = std::ifstream{schema_path};
-    if (!ifs) {
-        throw std::runtime_error("Failed to read schema file");
-    }
-
-    return json::parse(ifs);
-}
-
-void validate_index(const std::filesystem::path &schema_directory, std::ifstream &ifs) {
-    // **YUCK**: We have to read in the data with jsoncons here rather than reusing
-    // the nlohmann-json representation :-(
-    const auto index = json::parse(ifs);
-
-    // Load schema
-    auto ifs_schema = std::ifstream{schema_directory / "data_index.json"};
-    if (!ifs_schema) {
-        throw std::runtime_error("Failed to load schema");
-    }
-
-    const auto resolver = [&schema_directory](const auto &uri) {
-        return resolve_uri(uri, schema_directory);
-    };
-    const auto schema = jsonschema::make_json_schema(json::parse(ifs_schema), resolver);
-
-    // Perform validation
-    schema.validate(index);
-}
-} // namespace
-
-namespace hgps::data {
-DataManager::DataManager(std::filesystem::path root_directory, VerboseMode verbosity)
-    : root_{std::move(root_directory)}, verbosity_{verbosity} {
-    auto full_filename = root_ / "index.json";
+nlohmann::json read_input_files_from_directory(const std::filesystem::path &root_directory) {
+    auto full_filename = root_directory / "index.json";
     auto ifs = std::ifstream{full_filename};
     if (!ifs) {
         throw std::invalid_argument(
             fmt::format("File-based store, index file: '{}' not found.", full_filename.string()));
     }
 
-    index_ = nlohmann::json::parse(ifs);
+    auto index = nlohmann::json::parse(ifs);
 
-    if (!index_.contains("version")) {
+    if (!index.contains("version")) {
         throw std::runtime_error("File-based store, invalid definition missing schema version");
     }
 
-    auto version = index_["version"].get<int>();
+    auto version = index["version"].get<int>();
     if (version != 2) {
         throw std::runtime_error(fmt::format(
             "File-based store, index schema version: {} mismatch, supported: 2", version));
     }
 
     // Validate against schema
     ifs.seekg(0);
-    const auto schema_directory = get_program_directory() / "schemas" / "v1";
-    validate_index(schema_directory, ifs);
+    const auto schema_directory = hgps::get_program_directory() / "schemas" / "v1";
+    hgps::data::validate_index(schema_directory, ifs);
+
+    return index;
+}
+} // anonymous namespace
+
+namespace hgps::data {
+DataManager::DataManager(std::filesystem::path path, VerboseMode verbosity)
+    : verbosity_{verbosity} {
+    if (std::filesystem::is_regular_file(path)) {
+        // If it's a file, assume it's a zip file
+        root_ = create_temporary_directory();
+        extract_zip_file(path, root_);
+    } else {
+        // Otherwise it's a folder
```

Comment:
The same, cannot you check if this is a directory rather than assuming anything?

I would include the following branches in the following order:
- Is it a directory?
- Is it a ZIP file?
- If neither, there's an error.

Reply:
Done